From 7b93ed68da4d6ae6c65f996bc714eea00c8fa0bb Mon Sep 17 00:00:00 2001
From: Murali Karicheri <m-karicheri2@ti.com>
Date: Tue, 26 Sep 2017 10:14:36 -0400
Subject: [PATCH 82/86] net: prueth: add debug stats counters

During the packet loss investigation, additional debug counters were
added to root cause the issue. Make these part of the ethtool stats
so that it can be used for debugging and support of the driver.

Signed-off-by: Murali Karicheri <m-karicheri2@ti.com>
---
 drivers/net/ethernet/ti/hsr_prp_firmware.h |  8 ++++++++
 drivers/net/ethernet/ti/prueth.c           |  8 ++++++++
 drivers/net/ethernet/ti/prueth.h           | 10 ++++++++++
 3 files changed, 26 insertions(+)

diff --git a/drivers/net/ethernet/ti/hsr_prp_firmware.h b/drivers/net/ethernet/ti/hsr_prp_firmware.h
index b8c7b1d..b8df94a 100644
--- a/drivers/net/ethernet/ti/hsr_prp_firmware.h
+++ b/drivers/net/ethernet/ti/hsr_prp_firmware.h
@@ -214,6 +214,14 @@
 #define LRE_NODE_TABLE_LOOKUP_ERROR_A          (LRE_START + 112)
 #define LRE_NODE_TABLE_LOOKUP_ERROR_B          (LRE_START + 116)
 #define LRE_NODE_TABLE_FULL                    (LRE_START + 120)
+#define LRE_TOTAL_RX_A                         (LRE_START + 124)
+#define LRE_TOTAL_RX_B                         (LRE_START + 128)
+#define LRE_OVERFLOW_PRU0                      (LRE_START + 132)
+#define LRE_OVERFLOW_PRU1                      (LRE_START + 136)
+#define LRE_DD_PRU0                            (LRE_START + 140)
+#define LRE_DD_PRU1                            (LRE_START + 144)
+#define LRE_CNT_SUP_PRU0                       (LRE_START + 148)
+#define LRE_CNT_SUP_PRU1                       (LRE_START + 152)
 
 #define IEC62439_CONST_DUPLICATE_ACCEPT                 0x01
 #define IEC62439_CONST_DUPLICATE_DISCARD                0x02
diff --git a/drivers/net/ethernet/ti/prueth.c b/drivers/net/ethernet/ti/prueth.c
index 22fac6e..57136d7 100644
--- a/drivers/net/ethernet/ti/prueth.c
+++ b/drivers/net/ethernet/ti/prueth.c
@@ -3682,6 +3682,14 @@ static int emac_set_settings(struct net_device *ndev, struct ethtool_cmd *ecmd)
 	{"lreNtLookupErrA", PRUETH_LRE_STAT_OFS(node_table_lookup_error_a)},
 	{"lreNtLookupErrB", PRUETH_LRE_STAT_OFS(node_table_lookup_error_b)},
 	{"lreNodeTableFull", PRUETH_LRE_STAT_OFS(node_table_full)},
+	{"lreTotalRxA", PRUETH_LRE_STAT_OFS(lre_total_rx_a)},
+	{"lreTotalRxB", PRUETH_LRE_STAT_OFS(lre_total_rx_b)},
+	{"lreOverflowPru0", PRUETH_LRE_STAT_OFS(lre_overflow_pru0)},
+	{"lreOverflowPru1", PRUETH_LRE_STAT_OFS(lre_overflow_pru1)},
+	{"lreDDCountPru0", PRUETH_LRE_STAT_OFS(lre_cnt_dd_pru0)},
+	{"lreDDCountPru1", PRUETH_LRE_STAT_OFS(lre_cnt_dd_pru1)},
+	{"lreCntSupPru0", PRUETH_LRE_STAT_OFS(lre_cnt_sup_pru0)},
+	{"lreCntSupPru1", PRUETH_LRE_STAT_OFS(lre_cnt_sup_pru1)},
 };
 
 static int emac_get_sset_count(struct net_device *ndev, int stringset)
diff --git a/drivers/net/ethernet/ti/prueth.h b/drivers/net/ethernet/ti/prueth.h
index 0d6a0fb..5c7555a 100644
--- a/drivers/net/ethernet/ti/prueth.h
+++ b/drivers/net/ethernet/ti/prueth.h
@@ -239,6 +239,16 @@ struct lre_statistics {
 	u32 node_table_lookup_error_a;
 	u32 node_table_lookup_error_b;
 	u32 node_table_full;
+
+	/* additional debug counters */
+	u32 lre_total_rx_a; /* count of all frames received at port-A */
+	u32 lre_total_rx_b; /* count of all frames received at port-B */
+	u32 lre_overflow_pru0; /* count of overflow frames to host on PRU 0 */
+	u32 lre_overflow_pru1; /* count of overflow frames to host on PRU 1 */
+	u32 lre_cnt_dd_pru0; /* count of DD frames to host on PRU 0 */
+	u32 lre_cnt_dd_pru1; /* count of DD frames to host on PRU 1 */
+	u32 lre_cnt_sup_pru0; /* count of supervisor frames to host on PRU 0 */
+	u32 lre_cnt_sup_pru1; /* count of supervisor frames to host on PRU 1 */
 } __packed;
 
 struct prueth_hsr_prp_node {
-- 
1.9.1

